### DISCUSSION TITLE
Initialize Complex Typed Atom with Empty State?

### DISCUSSION THREAD

**Discussion by summitWT** [pmndrs/jotai#2256] [Q&A] [answered]
Hi,

So I'm confused to what would be the best practice here if I am trying to initialize an empty state for an atom with a complex object type. 

I have a `Phone` type with 10+ properties, some of them nested. I would like to initialize an empty `Phone` object like so:

```ts
export const deviceAtom = atom<Phone>({});
```

But TS throws an error, saying I must have all of the `Phone` properties initialized. This would be tedious, especially with complex types, so I was wondering if there is a `util` that I should be using or a pattern that I'm not understanding to help with this scenario.

I have tried something like

```ts
export const deviceAtom = atom<Phone | null>(null); // Can't set null to be an object later
export const deviceAtom = atom<Phone | undefined>(undefined); // Can't set undefined to be an object later
export const deviceAtom = atom<Phone | {}>({}); // This sort of works, see below
```

The best solution I have found so far is to make TS assertions when accessing properties in my UI components. 

```ts
//atoms.ts
export const deviceAtom = atom<Phone | {}>({});

export const selectedDeviceAtom = atom(
  async get => {
    return get(deviceAtom);
  },
  async (_get, set, newPhone: Phone) => {
    return set(deviceAtom, newPhone);
  },
);
```

```ts
//ui.tsx
const [selectedDevice, setSelectedDevice] = useAtom(selectedDeviceAtom);

return (
  <PhoneItem
    phone={phone}
    brand={brand}
    isSelected={(selectedDevice as Phone).phoneName === phone.phoneName} // this is the best I can do right now
    onSelectPhone={() => onSelectPhone(phone)}
  />
```

All of the examples I see usually involve simple atoms that start as a `number` or a `string`, but I haven't seen one that easily handles a custom `type` with many properties on it.

**Comment by dai-shi** [comment]
Atom values must be immutable, so my recommendation would be `atom<Phone | null>(null)` and use optional chaining.
If you prefer an empty object, `atom<Phone | Record<string, never>>({})` might work.

**Reply by summitWT** [reply, marked as answer]
Thanks for the reply.

I tried both ways you suggested and while they work to some degree, they did not bring type safety to my React components in an elegant way (optional chaining as you said, or casting as `Phone` type in the UI components like I was doing before). 

After a lot of back and forth, I finally landed on this as the preferred solution!

`export const deviceAtom = atom<Phone>({} as Phone);`

This brings type safety to the UI components and doesn't force me to create any initial values for my object. 

It ended up being much simpler than expected 😃
